### ISSUE TITLE
Word Alignment by Fine-tuning Embeddings on Parallel Corpora

### ISSUE THREAD

**Issue by pocca2048** [pocca2048/ML-paper-reading#19]
**Conference** : EACL 2021
**Link** : https://arxiv.org/pdf/2101.08231.pdf
**Github**: https://github.com/neulab/awesome-align
**Authors' Affiliation** : CMU
**TL;DR** : propose 1) PLM + finetune on parallel text (objective to improve alignment quality), and 2) methods to effectively extract alignments from these fine-tuned models.

**Summary** :
Importantly, we show that it is possible to train multilingual word aligners that can obtain robust performance even in zero-shot settings, making them a valuable tool that can be used out-ofthe-box with good performance over a wide variety of language pairs.

# 2 Methods

* Task: given X = (x1, ..., xn), Y = (y1, ..., ym), find A = {(xi, yj)}

## 2.1 Extracting Alignments from Embeddings
1. X와 Y를 mBERT에 넣어서 contextualized embedding을 뽑음. h_x, h_y. (i-th layer를 뽑는데, i는 hyper-parameter)
2. 1) probability simplexes and 2) optimal transport에 기반한 unidirectional alignment scores를 계산 
3. 이 alignment scores를 alignment matrices로 변환. reconcile alignments in the forward and backward directions

### Probability Simplex
similarity matrix S = hx dot hy^T -> Normalise (softmax, \alpha-entmax) -> := source-to-target alignment matrix

### Optimal Transport
* goal of optimal transport = to find a mapping that moves probability from one distribution to another, which can be used to find an optimal matching of similar words between two sequences.

$$ \sum_{i,j}  C(x_i, y_j) S_{xyij} $$ where $S_{xy}1_m = p_x$ and $S_{xy}1_n = p_y$.

### Extracting Bidirectional Alignments

$A = (S_{xy} > c) ∗ (S^T_{yx} > c)$

### Handling Subwords
To convert them to word alignments, we follow previous work (Sabet et al., 2020; Zenkel et al., 2020) and consider two words to be aligned if any of their subwords are aligned.

## 2.2 Fine-tuning Contextualized Embeddings for Word Alignment
* MLM
* TLM (Translation LM)
* Self-training Objective (SO) = $\sum_{i, j} A_{ij} \frac{1}{2} (\frac{S_{xyij}}{n} + \frac{S_{yxij}^T}{m} $
  * encourages words aligned in the first pass of alignment to have further closer contextualized representations.
  * align src->tgt & tgt -> src to be symmetric
  * 좀 이상. A를 계산하는 데에 이미 S가 들어가는데 이걸 몇 번이나 반복해서 사용하면 별로 좋지 않을 듯.
  * Sxy랑 Syx랑 transpose만 시킨 같은 게 아닌 것인가??
* Parallel Sentence Identification (PSI) = [CLS] -> 
  * output a prediction score. encourages the overall alignments of embeddings on both word and sentence level to be closer together.
* Consistency Optimization (CO) = $ \frac { tr (S_{xy}^T S_{yx} ) } {\min(m, n)} $
  * explicitly encourage the consistency between the two alignment matrices.
  * S를 여기서 한 번 더 사용..

Final objective L = LMLM + LTLM + LSO + LP SI + βLCO

loss가 너무 많아서 제대로 밸런싱이 안 되어 있을 듯.

# 3 Experiments
기존 method에 비해 성능이 뚜렷하게 좋다고 보기는 어려움. 내부에서도 좋은 거랑 안 좋은 거랑 차이가 큼.
